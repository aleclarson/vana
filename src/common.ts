/** Resolve object intersections */
export type Id<T> = { [P in keyof T]: T[P] }

/** Any property name */
export type AnyProp = keyof any

export type Except<T, U> = T extends U ? never : T

/** Used in place of `object` when necessary */
export type Dictionary<T> = {
  [key: string]: T | undefined
  [index: number]: T | undefined
}

/** Add `dispose` method to type T */
export type Disposable<T extends object = any> = Id<T & IDisposable>

/** Object with `dispose` method */
export interface IDisposable {
  dispose: () => void
}

/** `Object#hasOwnProperty` bound to `Function#call` */
export const has = Function.call.bind(Object.hasOwnProperty) as (
  obj: object,
  prop: AnyProp
) => boolean

/** Create a function that reuses a property descriptor */
export const PropertyDescriptor = (desc: object) => {
  const attr = has(desc, 'get') ? 'get' : 'value'
  return (obj: object, prop: AnyProp, value: any) => {
    desc[attr] = value
    Object.defineProperty(obj, prop, desc)
    desc[attr] = undefined
    return obj
  }
}

/** For defining a non-enumerable property */
export const definePrivate = PropertyDescriptor({ value: undefined })

/** Function that ignores its arguments and always returns nothing */
export const noop = Function.prototype as (...args: any[]) => void

export const getProto = Object.getPrototypeOf
export const isArray = Array.isArray

export function isObject(value: unknown): value is object {
  return value && typeof value === 'object'
}

export function isPlainObject(value: unknown): value is object {
  if (value) {
    const proto = getProto(value)
    return proto === null || proto === Object.prototype
  }
  return false
}

export function isCopyable(value: unknown) {
  return isArray(value) || isPlainObject(value)
}

export interface IThenable<T = any> {
  then(
    onFulfill: (value: T) => void,
    onReject?: (error: Error) => void
  ): IThenable<T>
}

export function isThenable(value: object): value is IThenable {
  return typeof value['then'] == 'function'
}

export function isGetter<T>(fn: Function): fn is () => T {
  return !fn.length
}

export function shallowCopy<T extends Dictionary<any>>(
  obj: T,
  exclude?: Array<keyof any>
): T {
  if (isArray(obj)) {
    return obj.slice() as any
  }
  let clone = Object.create(Object.getPrototypeOf(obj))
  each(obj, prop => {
    if (exclude && exclude.indexOf(prop) >= 0) return
    let desc = Object.getOwnPropertyDescriptor(obj, prop)!
    if (desc.get) {
      throw Error('Getters are only allowed on prototypes')
    }
    if (desc.enumerable) {
      clone[prop] = desc.value
    } else {
      Object.defineProperty(clone, prop, {
        value: desc.value,
        writable: true,
        configurable: true,
      })
    }
  })
  return clone
}

/** Iterate over the keys of an object */
export function each<T extends object>(obj: T, iter: (key: keyof T) => void) {
  if (isArray(obj)) {
    for (let i = 0; i < obj.length; i++) iter(i as any)
  } else if (typeof Reflect !== 'undefined') {
    Reflect.ownKeys(obj).forEach(iter as any)
  } else {
    Object.getOwnPropertyNames(obj).forEach(iter as any)
    Object.getOwnPropertySymbols(obj).forEach(iter as any)
  }
}
