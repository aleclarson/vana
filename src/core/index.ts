import { Observable } from './Observable'
import { Tapped } from './tap'

/** Useful for `subscribe` functions built on `vana` */
export type Observed<T> = T extends Observable<infer U> ? U : Tapped<T>

// Classes
export * from './Observable'
export * from './OBinding'
export * from './OPath'
export * from './OPromise'
export * from './OProps'
export * from './OTuple'
export * from './Change'

// Functions
export * from './array'
export * from './become'
export * from './commit'
export * from './commitArray'
export * from './commitIndices'
export * from './freeze'
export * from './keepAlive'
export * from './latest'
export * from './o'
export * from './revise'
export * from './tap'
export * from './watch'

// Immer
export { Draft, Immutable, isDraft, nothing, original, immerable } from 'immer'
export { Recipe, produce, setUseProxies, setAutoFreeze } from '../shared/immer'

// Shared
export { Disposable, IDisposable, IThenable, $O } from '../shared'
