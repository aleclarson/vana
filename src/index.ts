export { o } from './funcs/o'
export { tap, Tapped } from './funcs/tap'
export { watch, WatchProxy } from './funcs/watch'
export { revise } from './funcs/revise'
export { Domain } from './types/Domain'
export { Observable, isObservable, getObservable } from './types/Observable'

export {
  Draft,
  Recipe,
  Immutable,
  produce,
  nothing,
  isDraft,
  original,
  applyPatches,
  setUseProxies,
  setAutoFreeze,
} from './immer'

export { freeze, Frozen } from './funcs/freeze'
export { Disposable, IDisposable, IThenable } from './common'
