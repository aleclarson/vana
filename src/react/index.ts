export * from './useO'
export * from './useObserver'
export * from './useObserved'
export * from './useDerived'
export * from './useMemos'
export * from './useKeyedMemos'
export * from './useKeepAlive'
