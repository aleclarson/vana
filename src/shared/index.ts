export * from './symbols'

/** Give `any` its own class */
export abstract class Any {
  protected _: any
}

/** Resolve object intersections */
export type Id<T> = { [P in keyof T]: T[P] }

/** Any property name */
export type AnyProp = keyof any

/** Any array type */
export type AnyArray = ReadonlyArray<any>

/** Used in place of `object` when necessary */
export type Dictionary<T> = {
  [key: string]: T | undefined
  [index: number]: T | undefined
}

/** Add `dispose` method to type T */
export type Disposable<T extends object = any> = Id<T & IDisposable>

/** Object with `dispose` method */
export interface IDisposable {
  dispose: () => void
}

/** `Object#hasOwnProperty` bound to `Function#call` */
export const has = Function.call.bind(Object.hasOwnProperty) as (
  obj: object,
  prop: AnyProp
) => boolean

/** Create a function that reuses a property descriptor */
export const createDescriptor = (desc: object) => {
  const attr = has(desc, 'get') ? 'get' : 'value'
  return (obj: object, prop: AnyProp, value: any) => {
    desc[attr] = value
    Object.defineProperty(obj, prop, desc)
    desc[attr] = undefined
    return obj
  }
}

/** For defining a non-enumerable property */
export const definePrivate = createDescriptor({ value: undefined })

/** Function that ignores its arguments and always returns nothing */
export const noop = Function.prototype as (...args: any[]) => void

export const getProto = Object.getPrototypeOf
export const isArray = Array.isArray

export function isFunction(value: unknown): value is Function {
  return typeof value === 'function'
}

export function isObject(value: unknown): value is object {
  return value && typeof value === 'object'
}

export function isPlainObject(value: unknown): value is object {
  if (value) {
    const proto = getProto(value)
    return proto === null || proto === Object.prototype
  }
  return false
}

export interface IThenable<T = any> {
  then(
    onFulfill: (value: T) => void,
    onReject?: (error: Error) => void
  ): IThenable<T>
}

export function isThenable(value: object): value is IThenable {
  return typeof value['then'] == 'function'
}

export function shallowCopy<T extends object>(obj: T): T {
  if (isArray(obj)) {
    return obj.slice() as any
  } else {
    let copy = Object.create(getProto(obj))
    each(obj, prop => copyProp(obj, prop, copy))
    return copy
  }
}

export function copyProp(src: object, prop: AnyProp, dest: object) {
  let desc = Object.getOwnPropertyDescriptor(src, prop)!
  if (desc.get) {
    throw Error('Getters are only allowed on prototypes')
  }
  if (desc.enumerable) {
    dest[prop] = desc.value
  } else {
    Object.defineProperty(dest, prop, {
      value: desc.value,
      writable: true,
      configurable: true,
    })
  }
  return desc
}

/** Iterate over the keys of an object */
export function each(obj: object, iter: (prop: AnyProp) => void) {
  if (isArray(obj)) {
    for (let i = 0; i < obj.length; i++) iter(i)
  } else if (typeof Reflect !== 'undefined') {
    Reflect.ownKeys(obj).forEach(iter)
  } else {
    Object.getOwnPropertyNames(obj).forEach(iter)
    Object.getOwnPropertySymbols(obj).forEach(iter)
  }
}
