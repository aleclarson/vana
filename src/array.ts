import { AnyArray } from './common'
import { become } from './funcs/become'
import { commitArray } from './funcs/commitArray'
import { commitIndices } from './funcs/commitIndices'

export function append<T extends AnyArray>(
  base: T,
  first: T[number],
  ...rest: Array<T[number]>
): T

export function append(base: AnyArray, ...values: any[]) {
  const copy = base.slice()
  copy.push(...values)
  if (become(base, copy)) {
    commitIndices(base, copy, base.length, values.length)
    commitArray(base, copy)
  }
  return copy
}

export function prepend<T extends AnyArray>(
  base: T,
  first: T[number],
  ...rest: Array<T[number]>
): T

export function prepend(base: AnyArray, ...values: any[]) {
  const copy = values.concat(base)
  if (become(base, copy)) {
    commitIndices(base, copy, 0, copy.length)
    commitArray(base, copy)
  }
  return copy
}

export function insert<T extends AnyArray>(
  base: T,
  index: number,
  first: T[number],
  ...rest: Array<T[number]>
): T

export function insert(base: AnyArray, index: number, ...values: any[]) {
  while (index < 0) index += base.length
  const copy = base.slice(0, index).concat(values, base.slice(index))
  if (become(base, copy)) {
    commitIndices(base, copy, index, copy.length - index)
    commitArray(base, copy)
  }
  return copy
}

export function concat<T extends AnyArray>(
  base: T,
  first: Exclude<T[number], AnyArray> | T,
  ...rest: Array<typeof first>
): T

export function concat(base: AnyArray, ...values: any[]) {
  const copy = base.concat(...values)
  const delta = copy.length - base.length
  if (delta == 0) {
    return base
  }
  if (become(base, copy)) {
    commitIndices(base, copy, base.length, delta)
    commitArray(base, copy)
  }
  return copy
}

export function remove<T extends AnyArray>(
  base: T,
  index: number,
  count: number = 1
): T {
  while (index < 0) index += base.length
  if (count <= 0 || index >= base.length || index + count <= 0) {
    return base
  }

  const copy: any =
    index <= 0
      ? base.slice(index + count)
      : index + count >= base.length
      ? base.slice(0, index)
      : base.slice(0, index).concat(base.slice(index + count))

  if (become(base, copy)) {
    commitIndices(base, copy, index, base.length - index)
    commitArray(base, copy)
  }
  return copy
}
