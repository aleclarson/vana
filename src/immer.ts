import { Draft, Immer, IProduce } from 'immer'
import { definePrivate, isArray } from './common'
import { commit } from './funcs/commit'
import { __$observable, Observable } from './types/Observable'

/** Function that modifies a draft */
export type Recipe<
  T extends object = any,
  Args extends any[] = any[],
  Result = any
> = (draft: Draft<T>, ...args: Args) => Result

export const immer = new Immer({
  onAssign(state, prop: any) {
    state.assigned[prop] = true
  },
  onDelete(state, prop: any) {
    state.assigned[prop] = false
  },
  onCopy(state) {
    let { base } = state
    let target: Observable = base && base[__$observable]
    if (target) {
      let { copy, assigned } = state
      if (isArray(base)) {
        // Non-arrays have their __$observable property copied over by Immer.
        definePrivate(copy, __$observable, target)
      }
      target._rebind(copy)
      for (let prop in assigned) {
        commit(target, base[prop], copy[prop], prop, !assigned[prop], state)
      }
      commit(target, base, copy, null, false, state)
    }
  },
})

// tslint:disable:no-default-export
export default immer.produce

export const produce: IProduce = immer.produce
export const setAutoFreeze = immer.setAutoFreeze.bind(immer)
export const setUseProxies = immer.setUseProxies.bind(immer)
