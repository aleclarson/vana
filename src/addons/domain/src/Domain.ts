import { $O, Change, IObserver, o } from '../../../core'
import { Dictionary, isArray } from '../../../shared'

/**
 * The `Domain` class ensures every object it contains is always the latest revision.
 *
 * Domains are __mutable__.
 */
export class Domain<T extends object = any> implements IObserver<T> {
  private data: Dictionary<T>

  constructor(readonly key: string | number) {
    this.data = Object.create(null)
  }

  get(key: string | number): T | undefined {
    return this.data[key]
  }

  add(value: T) {
    let key = value[this.key]
    if (key == null) {
      throw Error(`The "${this.key}" property must exist`)
    }

    if (this.data[key]) {
      return false
    }

    let observable = value[$O] || o(value)[$O]
    if (!observable) {
      throw Error('This object has no observable type')
    }

    observable._addObserver(this)
    this.data[key] = value
    return true
  }

  delete(key: any) {
    let value = this.data[key]
    if (value) {
      delete this.data[key]
      value[$O]._removeObserver(this)
      return true
    }
    return false
  }

  /** Replace the existing set of objects. */
  reset(data: Dictionary<T> | T[]) {
    this.dispose()
    this.data = Object.create(null)
    if (isArray(data)) {
      data.forEach(value => this.add(value))
    } else {
      for (let key in data) this.add(data[key]!)
    }
  }

  /** Stop observing all objects within this domain. */
  dispose() {
    for (let key in this.data) {
      this.data[key]![$O]._removeObserver(this)
    }
  }

  /** @internal */
  ['_onChange'](change: Change<T>) {
    if (change.prop == null) {
      let key: any = change.newValue[this.key]
      if (key == null) {
        throw Error(`The "${this.key}" property must exist`)
      }

      if (key !== this.data[key]![this.key]) {
        throw Error(`The "${this.key}" property must be constant`)
      }

      this.data[key] = change.newValue
    }
  }
}
