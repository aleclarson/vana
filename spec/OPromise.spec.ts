import { o, tap } from '../src'

describe('new OPromise()', () => {
  it('observes fulfillment', () => {
    let p = o(defer())

    let spy = jest.fn()
    tap(p, spy)

    p.resolve('test')
    return p.then(() => {
      expect(spy).toBeCalled()
      expect(spy.mock.calls).toMatchSnapshot()
    })
  })

  it('observes rejection', () => {
    let p = o(defer())
    p.catch(() => {})

    let spy = jest.fn()
    tap(p, spy)

    p.reject('test')
    return p.catch(() => {
      expect(spy).toBeCalled()
      expect(spy.mock.calls).toMatchSnapshot()
    })
  })
})

interface Deferred<T> extends Promise<T> {
  resolve(value: T): void
  reject(error: any): void
}

function defer<T = any>(): Deferred<T> {
  let resolve: (_: any) => void
  let reject: (_: any) => void
  let promise: any = new Promise(($1, $2) => {
    resolve = $1
    reject = $2
  })
  promise.resolve = resolve!
  promise.reject = reject!
  return promise
}
