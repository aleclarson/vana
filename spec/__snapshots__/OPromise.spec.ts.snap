// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new OPromise() observes fulfillment 1`] = `
Array [
  Array [
    Object {
      "error": undefined,
      "resolved": true,
      "result": "test",
    },
  ],
]
`;

exports[`new OPromise() observes rejection 1`] = `
Array [
  Array [
    Object {
      "error": "test",
      "resolved": true,
      "result": undefined,
    },
  ],
]
`;
